### PR TITLE
chore: add js test with mocha and chai

### PR DESCRIPTION
**Pull request by glennr**
This commit adds a JS unit testing framework with mocha and chai.
Sample unit tests are provided for Comment and CommentList.

Fixes https://github.com/shakacode/react-webpack-rails-tutorial/issues/9

### CODE REVIEW

**Comment by justin808** [comment]
Wow! Thanks @glennr!

@alexfedoseev @robwise, let's chat later today about how this compares to what we were planning for our internal project.

Thanks again @glennr! 

@robwise JS tests will eventually go into the react_on_rails generator.

**Comment by glennr** [comment]
@justin808 you're welcome - I hope it's useful. Either way I'm super interested in knowing what test stack you guys end up going with on your internal project.

**Comment by glennr** [comment]
I found an issue on this branch: 

npm ERR! peerinvalid The package react@0.14.0 does not satisfy its siblings' peerDependencies requirements!
npm ERR! peerinvalid Peer react-bootstrap@0.27.1 wants react@>=0.14.0
npm ERR! peerinvalid Peer react-addons-test-utils@0.14.1 wants react@^0.14.1
npm ERR! peerinvalid Peer uncontrollable@3.1.3 wants react@>=0.11.0
npm ERR! peerinvalid Peer react-overlays@0.5.0 wants react@>=0.14.0

I'll push a fix for this.

**Comment by justin808** [comment]
@glennr I'm trying to run this, but it looks like a few more hours before we can update all.

```
  ~/shakacode/react-webpack-rails-tutorial/client (update-node-dependencies u=) ✗ npm install                                                                                                                                                                   ✹ [21:12:13]
- esprima-fb@15001.1001.0-dev-harmony-fb node_modules/defs/node_modules/esprima-fb
react-webpack-rails-tutorial@1.1.0 /Users/justin/shakacode/react-webpack-rails-tutorial/client
├─┬ UNMET PEER DEPENDENCY babel-core@6.0.14
│ ├── babel-code-frame@6.0.14
│ ├── babel-generator@6.0.14
│ ├── babel-helpers@6.0.14
│ ├── babel-messages@6.0.14
│ ├── babel-runtime@6.0.14
│ ├─┬ babel-template@6.0.14
│ │ └── babel-runtime@5.8.29
│ ├─┬ babel-traverse@6.0.14
│ │ └── babel-runtime@5.8.29
│ ├─┬ babel-types@6.0.14
│ │ └── babel-runtime@5.8.29
│ ├─┬ babylon@6.0.14
│ │ └── babel-runtime@5.8.29
│ └─┬ regenerator@0.8.35
│   ├── esprima-fb@15001.1.0-dev-harmony-fb
│   └─┬ recast@0.10.24
│     └── esprima-fb@15001.1.0-dev-harmony-fb
├─┬ eslint@1.8.0
│ └─┬ optionator@0.5.0
│   └── wordwrap@0.0.3
├─┬ jade@1.11.0
│ └─┬ uglify-js@2.4.24
│   └─┬ yargs@3.5.4
│     └── wordwrap@0.0.2
└─┬ react-bootstrap@0.27.3
  └── babel-runtime@5.8.29

npm WARN EPEERINVALID babel-loader@5.3.3 requires a peer of babel-core@^5.0.0 but none was installed.
```

**Comment by glennr** [comment]
@justin808 I indeed forgot to npm-shrinkwrap after adding those deps (I recalled something to that extent, but couldn't find it in the Readme - I found it on your blog entry this morning).

So there's a couple of issues on this branch now

  * [ ] incompatibility of react-addons-test-utils (try 0.14.0 not 0.14.1)
  * [ ] resolve babel incompatibility (after shrinkwrap)  

I'm on a terribad 3g connection at the moment, so npm install has been a 'special' experience for me all morning. I'll check back in another few hours when npm install has hopefully completed :-)

**Comment by glennr** [comment]
I had to upgrade react to 0.14.1 to be compatible with react-addons-test-utils. All specs pass OK.

I think a lot of my issues were to do with my npm version, which I upgraded;

```
glennr@Django client[gr_mocha_tests*] % npm -v
2.14.7
glennr@Django client[gr_mocha_tests*] % npm install npm -g
/usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
npm@3.3.10 /usr/local/lib/node_modules/npm
glennr@Django client[gr_mocha_tests*] % npm -v
3.3.10
```

@justin808 - WRT the babel incompatibility - I had to run 'npm prune --production' to remove all local dev deps before npm shrinkwrap would work. Is that right?

**Comment by glennr** [comment]
@justin808 - I've left the relevant unsquashed commits so its easier for you to review - but happy to squash them once you've given the  +1

File: .gitignore [hunk 1]
```diff
@@ -18,6 +18,7 @@
 .env
 node_modules
 npm-debug.log
+client/npm-debug.log*
```
Comment (robwise) [inline .gitignore]: We should have this in there regardless! Good catch!

Comment (glennr) [inline .gitignore]: Thanks! 

I just realized npm-debug.log* will suffice. All I wanted to do is ignore those timestamped npm-debug.log.201501...etc  files

File: client/test/components/CommentList_spec.jsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    expect(list[1].props.author).to.equal('Frank');
+  });
+
+  it('renders an alert if errors', () => {
+    const component = renderIntoDocument(
+      <CommentList $$comments={comments} error="zomg" />
+    );
+
+    const alert = findRenderedDOMComponentWithTag(component, 'strong');
+    expect(alert.textContent).to.equal('Comments could not be retrieved. ');
```
Comment (justin808) [inline client/test/components/CommentList_spec.jsx]: Wonder why there's an extra trailing space. Any idea?

File: client/test/test_helper.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+const doc = jsdom.jsdom('<!doctype html><html><body></body></html>');
+const win = doc.defaultView;
+
+global.document = doc;
+global.window = win;
+
+// from mocha-jsdom https://github.com/rstacruz/mocha-jsdom/blob/master/index.js#L80
+function propagateToGlobal(window) {
+  for (const key in window) {
+    if (!window.hasOwnProperty(key)) continue;
```
Comment (justin808) [inline client/test/test_helper.js]: this is OK. I'm perfectly happy if you want to use lodash and make this a bit more elegant. 

https://lodash.com/docs#forOwn

I'm of the opinion of using lodash more in this demo app.

**Comment by justin808** [comment]
CC: @alexfedoseev @glennr @jbhatab @robwise @mapreal19 @josiasds 

This looks great!

I say we merge it!

@glennr Go ahead and rebase to one commit with a nice message.

**Comment by samnang** [comment]
This looks great! :+1: 

Thank @glennr

**Comment by alexfedoseev** [comment]
And my :+1: 

Thanks @glennr!

**Comment by josiasds** [comment]
@glennr Great! :+1:

**Comment by glennr** [comment]
Awesome!

**Comment by justin808** [comment]
@glennr Can you please resolve the merge conflict and rebase again, and I'll merge.

Thanks again. Awesome job!

**Comment by justin808** [comment]
@glennr if you can take care of this, that would be great. I want to leave you in as the author.

**Comment by justin808** [comment]
@glennr I'm taking care of this now.

**Comment by justin808** [comment]
See #151

**Pull request closed without merging**